libs/sdk-ui-dashboard/src/model/store/catalog/catalogSelectors.ts

```diff
@@ -281,3 +293,58 @@ export const selectCatalogDateAttributeToDataset: DashboardSelector<
         capabilities.hasTypeScopedIdentifiers,
     );
 });
+
+function getAttributesWithHierarchyDescendants(
+    attributes: ICatalogAttribute[],
+    dateAttributes: ICatalogDateAttribute[],
+    attributeHierarchies: ICatalogAttributeHierarchy[],
+    ignoredDrillDownHierarchies: IDrillDownReference[] = [],
+): Record<string, ObjRef[]> {
+    const allCatalogAttributes = [...attributes, ...dateAttributes];
+    const attributeDescendants: Record<string, ObjRef[]> = {};
+
+    allCatalogAttributes.forEach((attribute) => {
+        const attributeRef = attribute.attribute.ref;
+        attributeHierarchies.forEach((hierarchy) => {
+            const hierarchyAttributes = hierarchy.attributeHierarchy.attributes.filter((attrRef) => {
+                const ignoredIndex = ignoredDrillDownHierarchies.findIndex((reference) => {
+                    if (isAttributeHierarchyReference(reference)) {
```

Comment:
Please extract this utility block somewhere. It is repeated multiple times in the pull code. (311-323 lines)

Reply:
Extracted to [attributeHierarchyUtils.ts](https://github.com/gooddata/gooddata-ui-sdk/pull/4339/files#diff-9f3c4d8d04ae077a880336f4b79fb262cd1867580d690411c91f460be18ee254) and replaced all duplicated code